Reportback

chloealee opened the pull request:
#### What does this PR do?

Allows the `$full` param to be passed so we can set to `false` when bulk-processing (will not make call to ns). 
#### Where should the reviewer start?

lib/modules/dosomething/dosomething_reportback/includes/Reportback.php line 142, then 162-175, then 127

lib/modules/dosomething/dosomething_reportback/includes/ReportbackItem.php line 95, then 120-133, then 80
#### How should this be manually tested?

When approve/promote/flag etc. reportback page is saved, northstar call is not run. Test with New Relic?
#### What are the relevant tickets?

Fixes #5415

chloealee committed:
adds shallowLoad param to Reportback:build and ReportbackItem:build

chloealee committed:
adds comment about user data

chloealee committed:
deletes extra line

chloealee committed:
changes  to  and adds param in fund function

chloealee committed:
adds  param to tell when to get full user data and make ns call

chloealee committed:
remove dies and print_rs

chloealee commented:
Hey @mshmsh5000 ! I think this is what you meant with the shallow load param but let me know if I'm totally off! 

Also, @weerd had a great suggestion to help with load time by putting all `drupal_id`s in an array and sending to ns all in one call rather than one at a time like I did here https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_northstar/dosomething_northstar.module#L128

mshmsh5000 commented:
The code looks like what I imagined!

Can you please remove `._dosomething_northstar.install` from this PR? I'm guessing it's some temp file created by your editor.

[Here](https://github.com/DoSomething/phoenix/pull/5408/files#diff-14a05db5e6c9435171679a1e613d23dfR140) and [here](https://github.com/DoSomething/phoenix/pull/5408/files#diff-14a05db5e6c9435171679a1e613d23dfR140), I'd specify that this param controls whether to load full _user_ data, since the rest of the Reportback/Item data load in both cases.

Re: The great suggestion from @weerd, depends on whether NS has an endpoint to accept lists of IDs -- I just don't know. On the Drupal side, to maintain backwards compatibility, you could implement like 

```
/**
 * Get user data from Northstar.
 * 
 * @param mixed $ids Single user ID, or array of user IDs
 * @return 
 */
function dosomething_northstar_get_northstar_user($id) {
    // Normalize to array
    $id = (!is_array($id)) ? array($id) : $id;

    // ... Loop through the jamz
}
```

Also note that because of #5415 we're never loading from cache. We'll have to fix that, too, as a core performance bug.

chloealee committed:
put dosomething_northstar.install file in lib/modules/dosomething/dosomething_northstar/

chloealee committed:
specifies user data in comment block

chloealee committed:
fixes cache logic to get user data from ns

chloealee commented:
@mshmsh5000 thanks for the comments! I don't know why those files got added and moved again but deleted `._dosomething_northstar.install` and moved `dosomething_northstar.install` to the correct place. Also specified user data in the comment blocks.

Re: suggestion from @weerd, it looks like you can only retrieve either one user or all users right now. Might be worth creating a separate issue for this to tackle?

Re: #5415 - added a fix and updated the PR description :)

mshmsh5000 commented:
Looks good. Re: #5415, here's my suggested version:

https://gist.github.com/mshmsh5000/1f011e2073c05ba00652

I've added exception logging for error responses from Northstar. We should try to catch and log those, so we know when Phoenix and Northstar aren't playing well together.

This will add Watchdog log entries like this:

```
Exception: Error fetching Northstar user data, uid=1704977: 'Not Found' in dosomething_northstar_get_northstar_user() (line 153 of/var/www/dev.dosomething.org/lib/modules/dosomething/dosomething_northstar/dosomething_northstar.module).
```

Let me know what you think!

chloealee committed:
adds exception logging for error responses from Northstar

chloealee commented:
Thanks for this, Matt! Makes total sense and very smart. Added your suggestion :)

weerd commented:
Howdy! Yep yep, NS doesn't support that yet. It was a suggestion that would require a bit of extra lift to add that capability to NS to help further efficiency in those requests. Uy created an issue for it today: https://github.com/DoSomething/northstar/issues/212

So one thing to note w/ the full/shallow load approach. This will affect all `find()` related calls at the moment. So any calls to the API that aim to find RBs or RB Items based on filters passed via params in the URL request to API will be affected. Just something to note. Otherwise need to add some sort of functionality (another URL param? something in class transformer?) to allow deciding _when_ to do a `shallow` load instead of the full load. Noting here in case it affects current API calls already retrieving user info that was bringing back full details from NS.

@aaronschachter @jonuy ^^

chloealee commented:
thanks, @weerd ! @aaronschachter @jonuy this should be fine though since right now, it's just getting one reportback item at a time, correct?

jonuy commented:
We actually do query for a bunch of reportback-items at a time. On the Android app I think it's set to 25 at a time in order to populate the gallery of photos on the main feed.

@chloealee Do you think you'd want to tackle DoSomething/northstar#212 as an additional task on top of this issue? Not sure what your workload is looking like right now. Otherwise, I could try to help by doing that one.

chloealee commented:
oooh gotcha, thanks @jonuy! Right now, I'm working on sending address from phoenix->ns since I messed that up when I tried to send over global country code stuff, [Enabling the retrieve endpoint for the taxonomy_term resource](https://github.com/DoSomething/phoenix/issues/5342), and [forwarding campaign activity to northstar](https://github.com/DoSomething/phoenix/issues/4571). I'm not sure if I can finish all by Friday but will try my best to! Priority wise, what should be tackle this first?

mshmsh5000 commented:
Sounds as though we need to resolve the find() problem that @weerd [mentions above](https://github.com/DoSomething/phoenix/pull/5408#issuecomment-146585864) before merging this. Or at least affirm that it won't screw up calls the native apps are making now.

Correct @aaronschachter @jonuy @chloealee?

chloealee commented:
From what I understand that makes sense! I'll get started on that. @aaronschachter @jonuy right now, we only want full load for the gallery of reportback items? Perhaps, per @weerd 's suggestion, I can add another URL param $gallery, and if that exists, change pass `true` instead of `false`?

weerd commented:
Cool! @chloe one suggestion would be to name the parameter something other than `$gallery`. That seems very specific to the case of loading the items for the gallery on mobile. But what if down the line we want to load the full data for some other purpose... passing a `?gallery=true` or something like that would seem odd and not make sense given a different context. I'd say name the parameter something that implies what it actually does which is indicate _retrieval of full user data_, so maybe a short word or two that could describe that? Not sure if I explained this clearly :grimacing:

chloealee commented:
@weerd super smart and totally makes sense - thanks for flagging! I'll name the parameter `$full_user` :)

weerd commented:
:+1:

aaronschachter commented:
So what should the updated `reportback-items` query look like in the mobile apps? `reportback-items?full_data=true` ?
`full_data` seems a little vague. Possibly something like `load_user_data`? `full_user_data` ?

jonuy commented:
I'll throw my vote for `?load_user_data=true` into the hat. :tophat:

chloealee commented:
@aaronschachter @jonuy you got it! `$load_user_data` it is.

chloealee committed:
adds load_user_data param

chloealee committed:
adds extra } that was ommitted by accident

mshmsh5000 commented:
This follows the "do the most obvious thing" rule, which is always +1.

CTO
@mshmsh5000 http://twitter.com/mshmsh5000 @dosomething
http://twitter.com/dosomething

On Wed, Oct 14, 2015 at 1:26 PM, Chloe Lee notifications@github.com wrote:

> @aaronschachter https://github.com/aaronschachter @jonuy
> https://github.com/jonuy you got it! $load_user_data it is.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/DoSomething/phoenix/pull/5408#issuecomment-148125756.

chloealee committed:
adds query

chloealee committed:
param works

chloealee committed:
gets rid of print_r and die comments

chloealee commented:
@weerd @mshmsh5000 added the `$load_user_data` param and confirms that it is working correctly! I was having trouble testing to make sure if it returned the ns user data because I think the cache already set to null but commented out the cache lines and worked fine.

weerd commented on lib/modules/dosomething/dosomething_reportback/includes/Reportback.php:
```diff
@@ -121,12 +121,22 @@ public static function find(array $filters = []) {
-      // @TODO: remove need for passing variable for constructor check.
-      $reportback = new static(['ignore' => TRUE]);
-      $reportback->build($item);
+    if (isset($filters['load_user_data'])) {
```
I don't think you need the `load_user_data` check here, which forces you to repeat the code in both the `if` and then the `else`. I would recommend putting it in the `build()` function. Then it's only a single check there when building out the full object... helps keep it DRY :smile: Instead of passing a hardcoded `true/false` to the `build()` function, you can now just pass the value of the `$filters['load_user_data']` which itself is now a boolean :)

``` php
<?php
$load_user_data = dosomething_helpers_isset($filters['load_user_data'], NULL, FALSE);
$reportback->build($item, $load_user_data);
```

It'll pass `true` or `false` to the build function without need for the `if/else` and duplicate code :dancer:

weerd replied:
the `dosomething_helpers_isset()` function is just a helper function to help deal w/ the whole `isset()` crap :)

chloealee replied:
oh this is so smart, thanks Diego! I was trying to figure out how to extract the code so I wouldn't repeat lines - so helpful!

weerd commented on lib/modules/dosomething/dosomething_reportback/includes/Reportback.php:
```diff
@@ -137,8 +147,9 @@ public static function find(array $filters = []) {
+   * @param bool  $full Boolean to decide whether to fetch full user data.
    */
-  private function build($data) {
+  private function build($data, $full = true) {
```
We should probs default this to `false` now that we have a URL parameter to control it, and since the resource file also defaults it to false in the services specification array.

chloealee replied:
changed!

chloealee committed:
extracts code to make more DRY

weerd commented:
Coool! Looking good :dancer: :+1:

chloealee commented:
sorry - just found another bug with the cache - it's not setting quite right anymore. I'll submit another comment for final approval when it's working again!

chloealee committed:
fixes cache in dosomething_northstar_get_northstar_user

chloealee commented:
@mshmsh5000 @weerd this should be the last one! Everything works as it should now :) Who do I need approval from to merge?

chloealee commented:
Oh right - I forgot about the undeclared variable part! That makes sense why it's working when I changed it back because it still makes the call to ns every time. 

I changed it back because right now, it returns the user info. the first time when it makes the call but when I refresh to test that `cache_get` is working properly, it returns `null` for all the user's data.  Found that if I delete the optional `$bin` parameter in `cache_set`, it now seems to work. 

One thing though - when I visit the site, I'm getting the error 

```
Warning: json_decode() expects parameter 1 to be string, object given in Reportback->build() (line 165 of /var/www/dev.dosomething.org/lib/modules/dosomething/dosomething_reportback/includes/Reportback.php).
```

but I'm pretty sure it is already passing it a string, especially because it worked the first time too. Is this a concern?

weerd commented:
@chloealee something must have changed that it's now passing an object. If it's popping up as an error message it's definitely a concern. We'd want to fix that! :wrench:

chloealee committed:
fixes cache

chloealee committed:
fixes cache

chloealee commented:
@mshmsh5000 @weerd thanks for your patience and help with this - finally figured out the issue and confirms that cache is working properly and no errors appear on the site anymore! 

@weerd I tried to change `load_user_data` to just `load_user` but for some reason, that broke everything. Looks like @aaronschachter also added the query string already - should we keep for now or still try to make the change?

weerd commented:
Might have been a cache clear thing again if it broke everything... but if it's too much lift if Aaron already added the query string on a released version of the app, then maybe hold off?

aaronschachter commented:
O Hai, the query string isn't in a release yet, so it's easy to change, just let me know what to do

weerd commented:
Cool! Then maybe give it a whirl @chloealee and just make sure to clear cache when you do, cause Droooops will be remembering the old infoz :wink:

chloealee committed:
changes param to load_user_data to load_user

chloealee commented:
@weerd always right! `drush cc all`: 2, Chloe: 0. Double fail for this week :-1: 

Just updated the param to `load_user` :) @aaronschachter

weerd commented:
@chloealee I had the same stats at first. Eventually you get used to the Drooops idiosyncrasies :wink:

weerd commented on lib/modules/dosomething/dosomething_northstar/dosomething_northstar.module:
```diff
@@ -125,20 +125,32 @@ function dosomething_northstar_update_user($form_state) {
+        $error = sprintf("Error fetching Northstar user data, uid=%d: '%s'", $id, $northstar_user_info->error);
+        watchdog_exception('northstar', new Exception($error));
+      } else {
+        // die($northstar_user_info);
```
remove debug stuffz :8ball:

weerd commented:
Just one debug item to remove, other than that looks good! :+1:

chloealee committed:
gets rid of debug line

chloealee commented:
gah - thanks for catching that! Removed :)

weerd commented:
:+1:

mshmsh5000 commented:
:+1:
